c10/core/ScalarType.h

```diff
@@ -46,15 +46,15 @@ namespace c10 {
   _(c10::quint8, QUInt8) /* 13 */                        \
   _(c10::qint32, QInt32) /* 14 */                        \
   _(at::BFloat16, BFloat16) /* 15 */                     \
-  _(c10::quint4x2, QUInt4x2) /* 16 */                    \
-  _(c10::quint2x4, QUInt2x4) /* 17 */                    \
-  _(c10::bits1x8, Bits1x8) /* 18 */                      \
-  _(c10::bits2x4, Bits2x4) /* 19 */                      \
-  _(c10::bits4x2, Bits4x2) /* 20 */                      \
-  _(c10::bits8, Bits8) /* 21 */                          \
-  _(c10::bits16, Bits16) /* 22 */                        \
-  _(c10::Float8_e5m2, Float8_e5m2) /* 23 */              \
-  _(c10::Float8_e4m3fn, Float8_e4m3fn) /* 24 */
+  _(c10::Float8_e5m2, Float8_e5m2) /* 16 */              \
```

Comment:
@ezyang can you confirm my understanding that this macro order is what associates int8 <-> scalar type values and so re-ordering will make use deserialize these dtypes into the wrong dtype.

@australopitek you can verify this by serializing an fp8 Tensor without this change and then try to load it after this change. If you geta Bits8 Tensor or a segfault, then it's not ok :D

Reply:
@albanD, simple check shown that the change of order doesn't break serialization/deserialization. My doubts were caused by the comment above this macro:
```
// NB: Order matters for this macro; it is relied upon in
// _promoteTypesLookup and the serialization format.
```
Anyway, I found there is no support for serialization of float8 dtypes, I haven't added them to the storage map in my original fp8 commit.

My proposal for this PR is to:
- add float8 dtypes to the storage map,
- verify more extensively that change of dtypes order doesn't break serialization,
- wait for @ezyang confirmation regarding above.

BTW, do you think this PR will be included in 2.1 release scope?

Reply:
I have pushed second commit:
- brought back the original datatypes order in macro, because I found some usage of it in the codebase,
- handled float8 promotion separately to avoid extending promotion table with Undefined rows/cols,
- regarding mentioned serialization of float8 dtypes, I couldn't find a way to handle them via torch.save/load. I have an error about missing key in _dtype_to_storage_type_map (storage.py), but the comment discourages from adding new dtypes to this map,
- anyway, I have added handlers to Storage class that I believe were missing. Probably support for float8 serialization should be clarified and handled in other PR/issue.